src/Instrumentation/Event.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * Class EventWithDuration.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\Instrumentation;
+
+/**
+ * A server-timing event.
+ *
+ * @package AmpProject\AmpWP
+ */
+class Event {
+
+	/**
+	 * Event name.
+	 *
+	 * @var string
+	 */
+	protected $name;
+
+	/**
+	 * Event description.
+	 *
+	 * @var string|null
+	 */
+	protected $description;
+
+	/**
+	 * Additional properties of the event.
+	 *
+	 * @var string[]
+	 */
+	protected $properties;
+
+	/**
+	 * Event constructor.
+	 *
+	 * @param string        $name        Event name.
+	 * @param string|null   $description Optional. Event description.
+	 * @param string[]|null $properties  Optional. Additional properties for the
+	 *                                   event.
+	 */
+	public function __construct( $name, $description = null, $properties = [] ) {
+		$this->name        = $name;
+		$this->description = $description;
+		$this->properties  = (array) $properties;
+	}
+
+	/**
+	 * Get the name of the event.
+	 *
+	 * @return string Event name.
+	 */
+	public function get_name() {
+		return $this->name;
+	}
+
+	/**
+	 * Get the description of the event.
+	 *
+	 * @return string Event description.
+	 */
+	public function get_description() {
+		return $this->description ?: '';
+	}
+
+	/**
+	 * Add additional properties to the event.
+	 *
+	 * @param string[] $properties Properties to add.
+	 */
+	public function add_properties( $properties ) {
+		$this->properties = array_merge(
+			(array) $properties,
+			$this->properties
+		);
+	}
+
+	/**
+	 * Get the server timing header string.
+	 *
+	 * @return string Server timing header string representing this event.
+	 */
+	public function get_header_string() {
+		$property_strings = [];
+
+		foreach ( $this->properties as $property => $value ) {
+			if ( is_float( $value ) ) {
+				$property_strings[] = sprintf(
+					';%s="%.1f"',
+					addslashes( $property ),
+					$value
+				);
+			} else {
+				$property_strings[] = sprintf(
+					';%s="%s"',
+					addslashes( $property ),
+					addslashes( $value )
```

Comment:
Why is `addslashes` necessary?

Reply:
I added `addslashes()` to avoid having values with quotes break strings when they rendered as HTML and sent as headers.

I don't think we have a case right now where we'd have quotes in the property name or value, but as this represents a form of API, we need to harden it for possible future uses as well.

Do you think there's a better way for doing so?

Reply:
No, I just hadn't seen header values being escaped with backslashes. But I can see this is spec'ed: https://tools.ietf.org/id/draft-nottingham-structured-headers-00.html#string